flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mongodb-cdc/src/main/java/org/apache/flink/cdc/connectors/mongodb/source/reader/fetch/MongoDBStreamFetchTask.java

```diff
@@ -108,7 +110,32 @@ public void execute(Context context) throws Exception {
         this.taskRunning = true;
         try {
             while (taskRunning) {
-                Optional<BsonDocument> next = Optional.ofNullable(changeStreamCursor.tryNext());
+                Optional<BsonDocument> next = Optional.empty();
+                try {
+                    next = Optional.ofNullable(changeStreamCursor.tryNext());
+                } catch (MongoQueryException e) {
+                    if (e.getErrorCode() == CHANGE_STREAM_FATAL_ERROR) {
+                        ChangeStreamOffset offset =
+                                new ChangeStreamOffset(streamSplit.getStartingOffset().getOffset());
+                        if (offset.getTimestamp() != null) {
+                            LOG.warn(
+                                    "Resume token {} has expired, fallback to timestamp startup mode.",
```

Comment:
Still initial startup mode  or latest mode, maybe "Resume token {} has expired, fallback to restart based on timestamp"?

 I'm not a mongodb expert. I wonder whether some data will loss or duplicated when fallback from resume token to timestamp? @Jiabao-Sun , CC

Reply:
Addressed that.